tpm/aziot-tpm-rs/src/tpm.rs

```diff
@@ -0,0 +1,177 @@
+// Copyright (c) Microsoft. All rights reserved.
+
+use std::convert::AsRef;
+use std::ops::{Deref, Drop};
+use std::os::raw::{c_uchar, c_void};
+use std::ptr;
+use std::slice;
+
+use aziot_tpm_sys::{
+    aziot_tpm_create, aziot_tpm_deinit, aziot_tpm_destroy, aziot_tpm_free_buffer,
+    aziot_tpm_get_keys, aziot_tpm_import_auth_key, aziot_tpm_init, aziot_tpm_sign_with_auth_key,
+    AZIOT_TPM_HANDLE,
+};
+
+use crate::Error;
+use crate::{ManageTpmKeys, SignWithTpm};
+
+/// Hsm for TPM
+///
+/// Create an instance of this to use the TPM interface of an HSM.
```

Comment:
There's only one "instance" of this implementation that needs to exist, so why does it need a handle? Why does it have to be `new()`'d ?

Eg this API makes it possible to call `new()` twice. What happens then?

Reply:
I mean, what would be the alternative? Having a single static global TPM handle instance in the C code which is "implicitly" passed to each function? I can't say I'm a huge fan of that...

Plus, while I'm all for deleting unused C code, I can't say I'm too excited to re-write the existing C code. The Rust-side bookkeeping required to store + eventually free a TPM handle is super straightforward, and doesn't make things all that more complicated.

As for what happens when `new()` is called twice? Well, based on the fact that this was also possible with the old API, and that the documentation never explicitly states this is a one-time global `new` function, I'm _assuming_ that nothing too bad would happen. i.e: it would simply return a fresh handle, and things would Just Work. That's a pretty naive view though, and I could be totally wrong.

If you've got any suggestions, I'm all ears, but as mentioned above, I'm of the opinion that if the code works, it'd be best to leave it as-is and focus on the rest of the task at hand.

Reply:
>As for what happens when `new()` is called twice? Well, based on the fact that this was also possible with the old API, and that the documentation never explicitly states this is a one-time global new function, I'm assuming that nothing too bad would happen. i.e: it would simply return a fresh handle, and things would Just Work. That's a pretty naive view though, and I could be totally wrong.

It's not quite as straightforward as reading the documentation. `initialize_tpm_device` internally mutates two statics to store the parameters for creating a new TPM connection. They're always updated in exactly the same way, and once the statics have been used to create a connection they can be mutated freely without causing problems to the already opened connection. So all it needs from you is to ensure that `aziot_tpm_init` cannot be called from two threads simultaneously. Of course you don't *need* to do that, because tpmd will only call `aziot_tpm_init` once, which is exactly my point.

You can keep `Tpm::new()` and have it return a handle, but ensure from the Rust side that calls to `aziot_tpm_init` are serialized.

To be safe it might be worth ensuring every `aziot_tpm_` call is serialized in general. The KS does this with `libaziot-keys` by having the bindings take `&mut self`, which eventually means the HTTP layer has a mutex to ensure only one request runs at a time, which is good enough. You can do that for the other fns of `Tpm` that currently take `&self`, but you can't do that for `Tpm::new()` itself, so you might need a `lazy_static` `Mutex` there.

>I'd like to avoid re-writing the existing C code as much as possible.

What *I'd* like to avoid is having all this C code of external provenance in the first place, but if we're going to have it I want to ensure it has the same thought put into it as if we were writing it ourselves.

Reply:
I'm going to assume that when you say `aziot_tpm_init` you actually mean `aziot_tpm_create`. `_init` doesn't do much aside from setup library-side logging. The `_create` method is the one that actually returns a new handle. 

> To be safe it might be worth ensuring every aziot_tpm_ call is serialized in general. The KS does this with libaziot-keys by having the bindings take &mut self, which eventually means the HTTP layer has a mutex to ensure only one request runs at a time, which is good enough. You can do that for the other fns of Tpm that currently take &self, but you can't do that for Tpm::new() itself, so you might need a lazy_static Mutex there.

Yeah, I thought that might be an issue, so I followed the same pattern you used in `aziot-keyd`. I'll be pushing those changes up soon-ish.

I'll also go ahead and add some sort of `lazy_static` / `std::sync::Once` based code to `aziot-tpm-rs` that ensures the methods you pointed out are only called once / serialized.

> What I'd like to avoid is having all this C code of external provenance in the first place, but if we're going to have it I want to ensure it has the same thought put into it as if we were writing it ourselves.

I feel that on a deep and visceral level, but based on my understanding of what you've said, the only benefit in changing the C API would be to remove the Rust-side handle book-keeping. The fundamental issue of ensuring certain methods are only called once / called by a single thread at a time won't be solved, right?

Reply:
>I'm going to assume that when you say `aziot_tpm_init` you actually mean `aziot_tpm_create`.

Right, sorry. I had written a paragraph about merging `_init` and `_create` into one to further reinforce that there is only one TPM connection, but lost it in editing.

>the only benefit in changing the C API would be to remove the Rust-side handle book-keeping. The fundamental issue of ensuring certain methods are only called once / called by a single thread at a time won't be solved, right?

For `_create` it would be solved by not having the Rust wrapper give the indication it could safely be called more than once, which it currently does by virtue of being called `new()` and returning a handle.

For the others, the lack of thread-safety is independent of whether there is a handle or not. It just requires something on the Rust side to ensure a `Mutex` is used somewhere. Having a handle does make it easier by making the handle methods take `&mut self` so that callers like the HTTP layer that have shared references are forced to use a `Mutex`.

Reply:
Okay, I feel like we're on the same page, and I'm just bad at formulating questions. Just to make sure: we can leave the C API as-is (i.e: using handles), and just tweak the Rust API to enforce the C API's soft invariants, yes?

Reply:
Yes.

Reply:
Addressed in 67fc30d